src/Analyser/MutatingScope.php

```diff
@@ -1246,6 +1246,33 @@ private function resolveType(Expr $node): Type
 			$leftType = $this->getType($left);
 			$rightType = $this->getType($right);
 
+			if ($leftType instanceof IntegerRangeType) {
+				$min = null;
+				$max = null;
+
+				if ($rightType instanceof ConstantIntegerType) {
+					if ($node instanceof Node\Expr\BinaryOp\Plus || $node instanceof Node\Expr\AssignOp\Plus) {
+						$min = $leftType->getMin() !== null ? $leftType->getMin() + $rightType->getValue() : null;
+						$max = $leftType->getMax() !== null ? $leftType->getMax() + $rightType->getValue() : null;
+					} elseif ($node instanceof Node\Expr\BinaryOp\Minus || $node instanceof Node\Expr\AssignOp\Minus) {
+						$min = $leftType->getMin() !== null ? $leftType->getMin() - $rightType->getValue() : null;
+						$max = $leftType->getMax() !== null ? $leftType->getMax() - $rightType->getValue() : null;
+					}
+				} elseif ($rightType instanceof IntegerRangeType) {
+					if ($node instanceof Node\Expr\BinaryOp\Plus || $node instanceof Node\Expr\AssignOp\Plus) {
+						$min = $leftType->getMin() !== null && $rightType->getMin() !== null ? $leftType->getMin() + $rightType->getMin() : null;
+						$max = $leftType->getMax() !== null && $rightType->getMax() !== null ? $leftType->getMax() + $rightType->getMax() : null;
+					} elseif ($node instanceof Node\Expr\BinaryOp\Minus || $node instanceof Node\Expr\AssignOp\Minus) {
+						$min = $leftType->getMin() !== null && $rightType->getMin() !== null ? $leftType->getMin() - $rightType->getMin() : null;
+						$max = $leftType->getMax() !== null && $rightType->getMax() !== null ? $leftType->getMax() - $rightType->getMax() : null;
+					}
+				}
+
+				if ($min || $max) {
+					return IntegerRangeType::fromInterval($min, $max);
+				}
+			}
+
 			$operatorSigil = null;
```

Comment:
below there is some `OperatorTypeSpecifyingExtension` magic.. should the new logic be placed in such a extension instead?